File: pi/src/flight_board/nearest.py
"""Distance ranking: haversine + top-N nearest, single pass."""

from __future__ import annotations

import math

from flight_board.source import Aircraft

EARTH_RADIUS_KM = 6371.0088


def haversine_km(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """Great-circle distance in kilometres between two lat/lon points."""
    rlat1, rlat2 = math.radians(lat1), math.radians(lat2)
    dlat = rlat2 - rlat1
    dlon = math.radians(lon2 - lon1)
    a = math.sin(dlat / 2) ** 2 + math.cos(rlat1) * math.cos(rlat2) * math.sin(dlon / 2) ** 2
    return 2 * EARTH_RADIUS_KM * math.asin(math.sqrt(a))


def bearing_degrees(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """Initial bearing from ``(lat1, lon1)`` to ``(lat2, lon2)`` in degrees."""
    rlat1, rlat2 = math.radians(lat1), math.radians(lat2)
    dlon = math.radians(lon2 - lon1)
    y = math.sin(dlon) * math.cos(rlat2)
    x = math.cos(rlat1) * math.sin(rlat2) - math.sin(rlat1) * math.cos(rlat2) * math.cos(dlon)
    return (math.degrees(math.atan2(y, x)) + 360.0) % 360.0


def nearest(
    aircraft: list[Aircraft],
    lat: float,
    lon: float,
    top_n: int,
) -> list[Aircraft]:
    """Return the ``top_n`` aircraft closest to ``(lat, lon)``.

    Distance is computed exactly once per aircraft (single pass), stored on
    ``dist_km``, and the sort key reuses that value rather than recomputing.
    """
    airborne = [ac for ac in aircraft if not ac.on_ground]
    for ac in airborne:
        ac.dist_km = haversine_km(lat, lon, ac.lat, ac.lon)
        ac.bearing_deg = bearing_degrees(lat, lon, ac.lat, ac.lon)
    ranked = sorted(airborne, key=lambda ac: ac.dist_km)
    if top_n < 0:
        return ranked
    return ranked[:top_n]

File: esp32/src/nearest.cpp
#include "nearest.h"

#include <algorithm>
#include <cmath>

static const double kEarthRadiusKm = 6371.0088;

double haversineKm(double lat1, double lon1, double lat2, double lon2) {
    double rlat1 = lat1 * M_PI / 180.0;
    double rlat2 = lat2 * M_PI / 180.0;
    double dlat = rlat2 - rlat1;
    double dlon = (lon2 - lon1) * M_PI / 180.0;
    double a = std::sin(dlat / 2) * std::sin(dlat / 2) +
               std::cos(rlat1) * std::cos(rlat2) * std::sin(dlon / 2) * std::sin(dlon / 2);
    return 2 * kEarthRadiusKm * std::asin(std::sqrt(a));
}

int radiusKmToNm(double km) {
    return (int)std::ceil(km / 1.852);
}

double bearingDegrees(double lat1, double lon1, double lat2, double lon2) {
    double rlat1 = lat1 * M_PI / 180.0;
    double rlat2 = lat2 * M_PI / 180.0;
    double dlon = (lon2 - lon1) * M_PI / 180.0;
    double y = std::sin(dlon) * std::cos(rlat2);
    double x = std::cos(rlat1) * std::sin(rlat2) -
               std::sin(rlat1) * std::cos(rlat2) * std::cos(dlon);
    return std::fmod(std::atan2(y, x) * 180.0 / M_PI + 360.0, 360.0);
}

void rankNearest(std::vector<Aircraft>& aircraft, double lat, double lon, int topN) {
    aircraft.erase(
        std::remove_if(aircraft.begin(), aircraft.end(),
                       [](const Aircraft& ac) { return ac.onGround; }),
        aircraft.end());
    for (Aircraft& ac : aircraft) {
        ac.distKm = haversineKm(lat, lon, ac.lat, ac.lon);
        ac.bearingDeg = bearingDegrees(lat, lon, ac.lat, ac.lon);
        ac.hasBearing = true;
    }
    std::sort(aircraft.begin(), aircraft.end(),
              [](const Aircraft& a, const Aircraft& b) { return a.distKm < b.distKm; });
    if (topN >= 0 && (int)aircraft.size() > topN) {
        aircraft.resize(topN);
    }
}
